### DISCUSSION TITLE
Should E3SM have a (almost) warning-free build?

### DISCUSSION THREAD

**Discussion by bartgol** [E3SM-Project/E3SM#5914] [Polls]
Building E3SM generates a _large_ amount of warnings. Some are linked to some compilers (e.g., some warnings from cuda compilers are "hard" to get rid of). However, there are a lot of warnings that are reproducible across compilers. They are often benign (e.g., unused variable), but contribute to make it impossible to spot a really concerning warning. In fact, I have long since given up looking at bldlog to see what warnings my code would add, since it would take me too long to parse it.

I am aware that fixing benign warnings may seem like a "style" task, and that "the model works, and we have science questions to answer". Alas, if the compiler warn you, it does so for a reason; even if 1 out of 1000 warnings is a _real_ bug, a code with vast downstream implications such as E3SM should probably guard against it.

**Comment by bartgol** [comment]
As I hinted in the post, I claim E3SM _should_ have a warning free build. Compiler warnings are often helpful to me, to spot issues ahead of time. As the number of warnings grows, people are more likely to ignore them, resulting in higher risk of bugs making their way into the main branch.

**Comment by rljacob** [comment]
Yes because this is the project that cares about good software practice and reducing warnings is part of that.   I think all developers should have this as a "background task" but we'd probably need some organized effort to make a dent in current warnings.

**Comment by mahf708** [comment]
The CI (at least public facing ones like circleCI) should fail with any warning imho

**Reply by bartgol** [reply]
There are _some_ warnings on _some_ machines/compilers that are hard to get around. I'm thinking about some Cuda warnings, for instance. There was one that I tried to "fix", only to give up and add a cudafe flag to ignore it (something related to a defaulted constructor or assignment operator). Of course, _if_ we find some cases like that, we can add some flag to ignore those warnings.

However, I agree that for some "easy enough" machine/compiler choice (like the ones you get with circleCI), we _should_ be able to address all warnings.

**Reply by rljacob** [reply]
I'd be happy with no CPU warnings.
